src/packages/utils/tests/logger.test.ts

```diff
@@ -0,0 +1,182 @@
+
+      try {
+        log(message);
+        await getCompletionHandle();
+      } finally {
+        closeSync(descriptor);
+        await unlink(path);
+      }
+
+      assert.strictEqual(
+        calls.length,
+        1,
+        "baseLogger() was called unexpected number of times."
+      );
+
+      const args = calls[0];
+
+      assert.deepStrictEqual(
+        args,
+        [message],
+        "baseLogger() called with unexpected arguments."
+      );
+    });
+
+    it("should write to the file provided", async () => {
+      const { descriptor, path } = getFileDescriptor("write-to-file-provided");
+      const { baseLogger } = createbaseLoggerger();
+
+      const { log, getCompletionHandle } = createLogger({
+        file: descriptor,
+        baseLogger
+      });
+
+      let fileContents: string;
+      try {
+        log(`${message} 0`);
+        log(`${message} 1`);
+        log(`${message} 2`);
+        await getCompletionHandle();
+
+        fileContents = await readFile(path, "utf8");
+      } finally {
+        closeSync(descriptor);
+        await unlink(path);
+      }
+
+      const logLines = fileContents.split("\n");
+
+      // 4, because there's a \n at the end of each line
+      assert.strictEqual(logLines.length, 4);
+      assert.strictEqual(logLines[3], "");
+
+      logLines.slice(0, 3).forEach((logLine, lineNumber) => {
+        const timestampPart = logLine.slice(0, 24);
+        const messagePart = logLine.slice(25);
+        const delimiter = logLine[24];
+
+        assert(timestampRegex.test(timestampPart), "Unexpected timestamp.");
+        assert.strictEqual(delimiter, " ", "Unexpected delimiter.");
+        assert.strictEqual(
+          messagePart,
+          `${message} ${lineNumber}`,
+          "Unexpected message."
+        );
+      });
+    });
+
+    it("should timestamp each line on multi-line log messages", async () => {
+      const { descriptor, path } = getFileDescriptor("timestamp-each-line");
+      const { baseLogger } = createbaseLoggerger();
+
+      const { log, getCompletionHandle } = createLogger({
+        file: descriptor,
+        baseLogger
+      });
+
+      const expectedLines = ["multi", "line", "message"];
+
+      let loggedLines: string[];
+      try {
+        log(expectedLines.join("\n"));
+        await getCompletionHandle();
+
+        const fileContents = await readFile(path, "utf8");
+        loggedLines = fileContents.split("\n");
+      } finally {
+        closeSync(descriptor);
+        await unlink(path);
+      }
+
+      // length == 4, because there's a \n at the end (string.split() results
+      // in each log line, follwed by an empty string)
+      assert.strictEqual(loggedLines.length, 4);
+      assert.strictEqual(loggedLines[3], "");
+
+      loggedLines.slice(0, 3).forEach((logLine, lineNumber) => {
+        const timestampPart = logLine.slice(0, 24);
+        const messagePart = logLine.slice(25);
+        const delimiter = logLine[24];
```

Comment:
What is this [kind of magic](https://www.youtube.com/watch?v=0p_1QSUsbsM)?

Reply:
👑 

Done!